pkg/controllers/manifests/init.go

```diff
@@ -96,9 +148,397 @@ func (r *InitManifestsConfigMapReconciler) Reconcile(ctx context.Context, req ct
+
+	exists, err := r.releaseExists(ctx, cm)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if exists {
+		r.Log.Infof("release for chart %s already exists", chartName)
+		// check if upgrade is needed
+		upgradedNeeded, err := r.checkIfUpgradeNeeded(ctx, cm)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		if upgradedNeeded {
+			// initiate upgrade
+			err = r.initiateUpgrade(ctx, cm)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+
+			// register this configuration
+			err = r.registerLastAppliedChartConfig(ctx, cm)
+			if err != nil {
+				r.Log.Errorf("error updating config map with last applied chart annotation: %v", err)
+				return ctrl.Result{}, err
+			}
+
+			return ctrl.Result{}, nil
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	r.Log.Infof("initaiting installation for chart %s", chartName)
+	// initiate install
+	err = r.initiateInstall(ctx, cm)
+	if err != nil {
+		r.Log.Errorf("error installing chart %s", chartName)
+		return ctrl.Result{}, err
+	}
+
+	// register this configuration
+	err = r.registerLastAppliedChartConfig(ctx, cm)
+	if err != nil {
+		r.Log.Errorf("error updating config map with last applied chart annotation: %v", err)
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *InitManifestsConfigMapReconciler) checkIfUpgradeNeeded(ctx context.Context, cm *corev1.ConfigMap) (bool, error) {
+	chartName := cm.Data["name"]
+
+	// compare with last applied config
+	lastAppliedConfig, ok := cm.ObjectMeta.Annotations[LastAppliedChartConfig]
+	if !ok {
+		// should register current config as the last applied config?
+		r.Log.Infof("release for chart %s exists but no last applied config exists", chartName)
+
+		err := r.registerLastAppliedChartConfig(ctx, cm)
+		if err != nil {
+			r.Log.Errorf("error updating config map with last applied chart annotation: %v", err)
+			return false, err
+		}
+
+		r.Log.Infof("updated current config as last applied config for chart: %s", chartName)
+		return false, nil
+	}
+
+	currentConfig, err := r.getCompressedConfig(ctx, cm)
+	if err != nil {
+		r.Log.Errorf("cannot get current config")
+		return false, err
+	}
+
+	if lastAppliedConfig != currentConfig {
+		// initiate upgrade
+		r.Log.Infof("current config different than last config, upgrade need for chart %s", chartName)
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func (r *InitManifestsConfigMapReconciler) initiateUpgrade(ctx context.Context, cm *corev1.ConfigMap) error {
+	chartName := cm.Data["name"]
+
+	lastAppliedConfig := cm.ObjectMeta.Annotations[LastAppliedChartConfig]
+	// first check if its the chart version that has changed
+	// as that would mean we first need to trigger a new archive to be pulled
+	lastConfigRaw, err := compress.Uncompress(lastAppliedConfig)
+	if err != nil {
+		r.Log.Errorf("unable to uncompress last applied config: %v", err)
+	}
+
+	lastConfig := make(map[string]string)
+	err = json.Unmarshal([]byte(lastConfigRaw), &lastConfig)
```

Comment:
I guess its better to use bundle hash here only as the bundle is already compressed and would make up quite a big portion of the annotation here